delegate-app/Main.hs

```diff
@@ -5,45 +5,57 @@ import Prelude
 
 -- Haskell imports
 
-import Control.Concurrent (threadDelay)
+import Control.Concurrent (Chan, newChan, readChan, threadDelay, writeChan)
+import Control.Concurrent.Async (async)
+import Control.Monad (forever, void)
 import Control.Monad.IO.Class (MonadIO (liftIO))
 
+-- Plutus imports
+import Plutus.V2.Ledger.Api (POSIXTime (..))
+
 -- HydraAuction imports
+
 import HydraAuction.Delegate (
-  DelegateEvent (Start),
+  DelegateEvent (..),
   DelegateInput (DelegateEvent),
   DelegateRunnerT,
   delegateStep,
   execDelegateRunnerT,
  )
-
-eventsProducer :: Bool -> DelegateRunnerT IO (Maybe DelegateEvent)
-eventsProducer thisIsFirstRun =
-  if thisIsFirstRun
-    then return $ Just Start
-    else return Nothing
-
-consumer :: Maybe DelegateInput -> DelegateRunnerT IO ()
-consumer mInput = case mInput of
-  Just input -> do
-    delegateResponse <- delegateStep input
-    -- FIXME: send response to client
-    liftIO $
-      putStrLn $
-        "Delegate responses for input: " <> show delegateResponse
-  Nothing -> return ()
-
-delegateTick :: Int
-delegateTick = 1_000
+import HydraAuction.Delegate.Interface (DelegateResponse (..))
+import HydraAuction.OnChain.Common (secondsLeftInInterval, stageToInterval)
+import HydraAuction.Tx.Common (currentAuctionStage, currentTimeMilliseconds)
+
+consumer :: Chan DelegateInput -> DelegateRunnerT IO ()
+consumer delegateInputs = do
+  -- This will block until an event appears in the queue
+  delegateIn <- liftIO $ readChan delegateInputs
```

Comment:
I do not understand what is wrong with one step. Like is not the purpose of queue exactly to process inputs one by one?

Reply:
conceptually, I think we currently run some action after each frontendrequest that can potentially need some amount of time, while that happens, the queue piles up, I'm just wondering whether, if we don't spawn specific handler threads, we're gonna have more `FrontendRequest`s per time than DelegateSteps per time.

Reply:
I think we wont run in this issue with the demo, but it might definitely be a problem for a real world application.
I propose we stick to a single reader/multiple writers for now.. it might turn out that relaxing to multiple readers is trivial, and we can do that at the end (or it might turn out to be hard, and we can ignore it 😄 )

Reply:
> if we don't spawn specific handler threads

I do not think that will help, cuz Delegate server does not do anything heavy, it just creates requests to Hydra node. So that is essentially non-blocking and non-IO-heavy task, which means that should get no gain by adding thread.

> but it might definitely be a problem for a real world application

I do not think where should be such situation even in production. We have 1000 ticks a second now. So if one bid requires like 10 inputs which seems like very pessimistic count to me, we can support 100 clients bidding every second. This seems like a lot to me.

Reply:
I think the way I've done in in my PR is perhaps a solution? Can you look at that?